File: packages/android/feedvalue/src/test/kotlin/com/feedvalue/sdk/FeedValueTests.kt
package com.feedvalue.sdk

import org.junit.Assert.*
import org.junit.Test

class FeedValueTests {

    // MARK: - FeedValueState Tests

    @Test
    fun `FeedValueState has correct default values`() {
        val state = FeedValueState()
        assertFalse(state.isReady)
        assertNull(state.config)
    }

    @Test
    fun `FeedValueState equality works correctly`() {
        val state1 = FeedValueState(isReady = true, config = null)
        val state2 = FeedValueState(isReady = true, config = null)
        assertEquals(state1, state2)
    }

    // MARK: - WidgetConfig Tests

    @Test
    fun `WidgetConfig stores all fields correctly`() {
        val config = WidgetConfig(
            widgetId = "test-123",
            name = "Test Widget",
            promptText = "How can we help?",
            thankYouMessage = "Thanks!"
        )

        assertEquals("test-123", config.widgetId)
        assertEquals("Test Widget", config.name)
        assertEquals("How can we help?", config.promptText)
        assertEquals("Thanks!", config.thankYouMessage)
    }

    @Test
    fun `WidgetConfig handles null optionals`() {
        val config = WidgetConfig(
            widgetId = "test-123",
            name = "Test Widget"
        )

        assertEquals("test-123", config.widgetId)
        assertEquals("Test Widget", config.name)
        assertNull(config.promptText)
        assertNull(config.thankYouMessage)
    }

    // MARK: - Feedback Tests

    @Test
    fun `Feedback creation with minimal fields`() {
        val feedback = Feedback(message = "Great app!")
        assertEquals("Great app!", feedback.message)
        assertNull(feedback.sentiment)
        assertNull(feedback.email)
        assertNull(feedback.metadata)
    }

    @Test
    fun `Feedback creation with all fields`() {
        val feedback = Feedback(
            message = "Love it!",
            sentiment = Sentiment.EXCITED,
            email = "user@example.com",
            metadata = mapOf("source" to "android")
        )

        assertEquals("Love it!", feedback.message)
        assertEquals(Sentiment.EXCITED, feedback.sentiment)
        assertEquals("user@example.com", feedback.email)
        assertEquals("android", feedback.metadata?.get("source"))
    }

    // MARK: - Sentiment Tests

    @Test
    fun `Sentiment has correct raw values`() {
        assertEquals("frustrated", Sentiment.FRUSTRATED.value)
        assertEquals("confused", Sentiment.CONFUSED.value)
        assertEquals("neutral", Sentiment.NEUTRAL.value)
        assertEquals("satisfied", Sentiment.SATISFIED.value)
        assertEquals("excited", Sentiment.EXCITED.value)
    }

    // MARK: - FeedValueError Tests

    @Test
    fun `NotInitialized error has correct message`() {
        val error = FeedValueError.NotInitialized
        assertTrue(error.message?.contains("not been initialized") == true)
    }

    @Test
    fun `NetworkError wraps underlying exception`() {
        val underlying = RuntimeException("Connection failed")
        val error = FeedValueError.NetworkError(underlying)
        assertTrue(error.message?.contains("Network error") == true)
        assertTrue(error.message?.contains("Connection failed") == true)
    }

    @Test
    fun `ServerError includes status code and message`() {
        val error = FeedValueError.ServerError(500, "Internal error")
        assertTrue(error.message?.contains("500") == true)
        assertTrue(error.message?.contains("Internal error") == true)
    }

    @Test
    fun `ServerError handles null message`() {
        val error = FeedValueError.ServerError(404, null)
        assertTrue(error.message?.contains("404") == true)
    }

    // MARK: - FeedValue Instance Tests

    @Test
    fun `getInstance returns null before initialize`() {
        val instance = FeedValue.getInstance("nonexistent-widget-id")
        assertNull(instance)
    }
}

// MARK: - API Response Parsing Tests

class FeedValueAPITests {

    @Test
    fun `ConfigResponse parsing works correctly`() {
        // Test data class creation (actual JSON parsing tested in integration)
        val config = WidgetConfigData(
            name = "Test Widget",
            promptText = "What do you think?",
            thankYouMessage = "Thank you!"
        )

        val response = ConfigResponse(
            widgetId = "widget-123",
            config = config,
            submissionToken = "token-abc",
            tokenExpiresAt = 1735689600L
        )

        assertEquals("widget-123", response.widgetId)
        assertEquals("Test Widget", response.config.name)
        assertEquals("token-abc", response.submissionToken)
        assertEquals(1735689600L, response.tokenExpiresAt)
    }

    @Test
    fun `ConfigResponse handles null optionals`() {
        val config = WidgetConfigData(
            name = "Test Widget",
            promptText = null,
            thankYouMessage = null
        )

        val response = ConfigResponse(
            widgetId = "widget-123",
            config = config,
            submissionToken = null,
            tokenExpiresAt = null
        )

        assertNull(response.submissionToken)
        assertNull(response.tokenExpiresAt)
        assertNull(response.config.promptText)
    }
}

File: packages/ios/Tests/FeedValueTests/FeedValueTests.swift
import XCTest
import Combine
@testable import FeedValue

final class FeedValueTests: XCTestCase {

    // MARK: - FeedValueState Tests

    func testFeedValueStateDefaultValues() {
        let state = FeedValueState()
        XCTAssertFalse(state.isReady)
        XCTAssertNil(state.config)
    }

    func testFeedValueStateEquality() {
        let state1 = FeedValueState(isReady: true, config: nil)
        let state2 = FeedValueState(isReady: true, config: nil)
        XCTAssertEqual(state1, state2)
    }

    // MARK: - WidgetConfig Tests

    func testWidgetConfigDecoding() throws {
        let json = """
        {
            "widget_id": "test-123",
            "name": "Test Widget",
            "prompt_text": "How can we help?",
            "thank_you_message": "Thanks!"
        }
        """
        let data = json.data(using: .utf8)!
        let config = try JSONDecoder().decode(WidgetConfig.self, from: data)

        XCTAssertEqual(config.widgetId, "test-123")
        XCTAssertEqual(config.name, "Test Widget")
        XCTAssertEqual(config.promptText, "How can we help?")
        XCTAssertEqual(config.thankYouMessage, "Thanks!")
    }

    func testWidgetConfigDecodingWithNilOptionals() throws {
        let json = """
        {
            "widget_id": "test-123",
            "name": "Test Widget"
        }
        """
        let data = json.data(using: .utf8)!
        let config = try JSONDecoder().decode(WidgetConfig.self, from: data)

        XCTAssertEqual(config.widgetId, "test-123")
        XCTAssertEqual(config.name, "Test Widget")
        XCTAssertNil(config.promptText)
        XCTAssertNil(config.thankYouMessage)
    }

    // MARK: - Feedback Tests

    func testFeedbackCreation() {
        let feedback = Feedback(message: "Great app!")
        XCTAssertEqual(feedback.message, "Great app!")
        XCTAssertNil(feedback.sentiment)
        XCTAssertNil(feedback.email)
        XCTAssertNil(feedback.metadata)
    }

    func testFeedbackWithAllFields() {
        let feedback = Feedback(
            message: "Love it!",
            sentiment: .excited,
            email: "user@example.com",
            metadata: ["source": "ios"]
        )
        XCTAssertEqual(feedback.message, "Love it!")
        XCTAssertEqual(feedback.sentiment, .excited)
        XCTAssertEqual(feedback.email, "user@example.com")
        XCTAssertEqual(feedback.metadata?["source"], "ios")
    }

    // MARK: - Sentiment Tests

    func testSentimentRawValues() {
        XCTAssertEqual(Sentiment.frustrated.rawValue, "frustrated")
        XCTAssertEqual(Sentiment.confused.rawValue, "confused")
        XCTAssertEqual(Sentiment.neutral.rawValue, "neutral")
        XCTAssertEqual(Sentiment.satisfied.rawValue, "satisfied")
        XCTAssertEqual(Sentiment.excited.rawValue, "excited")
    }

    // MARK: - FeedValueError Tests

    func testErrorDescriptions() {
        let notInitError = FeedValueError.notInitialized
        XCTAssertTrue(notInitError.localizedDescription.contains("not been initialized"))

        let networkError = FeedValueError.networkError(underlying: URLError(.notConnectedToInternet))
        XCTAssertTrue(networkError.localizedDescription.contains("Network error"))

        let serverError = FeedValueError.serverError(statusCode: 500, message: "Internal error")
        XCTAssertTrue(serverError.localizedDescription.contains("500"))
        XCTAssertTrue(serverError.localizedDescription.contains("Internal error"))

        let serverErrorNoMessage = FeedValueError.serverError(statusCode: 404, message: nil)
        XCTAssertTrue(serverErrorNoMessage.localizedDescription.contains("404"))
    }

    // MARK: - FeedValue Instance Tests

    func testGetInstanceReturnsNilBeforeInitialize() {
        let instance = FeedValue.getInstance(widgetId: "nonexistent-widget-id")
        XCTAssertNil(instance)
    }
}

// MARK: - Mock URL Protocol for API Tests

final class MockURLProtocol: URLProtocol {
    static var requestHandler: ((URLRequest) throws -> (HTTPURLResponse, Data))?

    override class func canInit(with request: URLRequest) -> Bool {
        return true
    }

    override class func canonicalRequest(for request: URLRequest) -> URLRequest {
        return request
    }

    override func startLoading() {
        guard let handler = MockURLProtocol.requestHandler else {
            XCTFail("No request handler set")
            return
        }

        do {
            let (response, data) = try handler(request)
            client?.urlProtocol(self, didReceive: response, cacheStoragePolicy: .notAllowed)
            client?.urlProtocol(self, didLoad: data)
            client?.urlProtocolDidFinishLoading(self)
        } catch {
            client?.urlProtocol(self, didFailWithError: error)
        }
    }

    override func stopLoading() {}
}

// MARK: - API Tests

final class FeedValueAPITests: XCTestCase {

    func testConfigResponseDecoding() throws {
        let json = """
        {
            "widget_id": "widget-123",
            "config": {
                "name": "Test Widget",
                "prompt_text": "What do you think?",
                "thank_you_message": "Thank you!"
            },
            "submission_token": "token-abc",
            "token_expires_at": 1735689600
        }
        """
        let data = json.data(using: .utf8)!
        let response = try JSONDecoder().decode(ConfigResponse.self, from: data)

        XCTAssertEqual(response.widgetId, "widget-123")
        XCTAssertEqual(response.config.name, "Test Widget")
        XCTAssertEqual(response.config.promptText, "What do you think?")
        XCTAssertEqual(response.submissionToken, "token-abc")
        XCTAssertEqual(response.tokenExpiresAt, 1735689600)
    }

    func testFeedbackResponseDecoding() throws {
        let json = """
        {
            "feedback_id": "fb-123",
            "success": true,
            "blocked": false,
            "message": "Feedback submitted"
        }
        """
        let data = json.data(using: .utf8)!
        let response = try JSONDecoder().decode(FeedbackResponse.self, from: data)

        XCTAssertEqual(response.feedbackId, "fb-123")
        XCTAssertEqual(response.success, true)
        XCTAssertEqual(response.blocked, false)
        XCTAssertEqual(response.message, "Feedback submitted")
    }

    func testErrorResponseDecoding() throws {
        let json = """
        {
            "detail": "Widget not found",
            "message": null,
            "error": null
        }
        """
        let data = json.data(using: .utf8)!
        let response = try JSONDecoder().decode(ErrorResponse.self, from: data)

        XCTAssertEqual(response.detail, "Widget not found")
        XCTAssertNil(response.message)
    }
}

// MARK: - DeviceIdentifier Tests

final class DeviceIdentifierTests: XCTestCase {

    func testFingerprintConsistency() {
        let identifier = DeviceIdentifier()

        let fingerprint1 = identifier.getOrCreateFingerprint()
        let fingerprint2 = identifier.getOrCreateFingerprint()

        // Same instance should return same fingerprint
        XCTAssertEqual(fingerprint1, fingerprint2)

        // Fingerprint should be 64 characters (SHA-256 hex)
        XCTAssertEqual(fingerprint1.count, 64)
    }

    func testFingerprintFormat() {
        let identifier = DeviceIdentifier()
        let fingerprint = identifier.getOrCreateFingerprint()

        // Should be valid hex string
        let hexCharacterSet = CharacterSet(charactersIn: "0123456789abcdef")
        XCTAssertTrue(fingerprint.unicodeScalars.allSatisfy { hexCharacterSet.contains($0) })
    }
}
